### PR TITLE
[stable/datadog] Add config maps, env vars, and ports

### PR DESCRIPTION
**Pull request by flah00**
* Optionally open up APM tracing ports
* Provide configmap hooks for custom autoconf, checks.d, and conf.d
* Open up environment variables

### CODE REVIEW

**Comment by flah00** [comment]
@gtaylor I feel like I should explain some of the choices I've made, regarding conf.d and auto_conf. 

The datadog entrypoint script recursively copies from `/conf.d` and `/checks.d` to `/etc/dd-agent/conf.d` and `/etc/dd-agent/checks.d`, respectively. If you aren't using k8s configmaps, you can easily create static and auto_conf configuration files. But the directories `/conf.d` and `/checks.d` are docker volumes. And confimaps do not permit sub-directories, at least as key names. Additional volumes cannot be mounted within these volumes (at least I haven't been able to figure out how). 

Thus, I have separated all three concerns. The `datadog.confd` and `datadog.checksd` keys write to their `/conf.d` and `/checks.d` counterparts. The `datadog.autoconf` (so named, because I can't use `_` or uppercase, otherwise k8s pukes) overwrites the existing `/etc/dd-agent/conf.d/autoconf` directory. This will effectively remove all files already present, such as etcd.yaml, redisdb.yaml, etc.

**Comment by lachie83** [comment]
cc @gtaylor

**Comment by seanknox** [comment]
Splitting out into distinct config maps makes sense to me. LGTM, but let's wait on @gtaylor to weigh in.

**Comment by gtaylor** [comment]
Sorry for the delay. This looks great. Much more flexible. LGTM.

**Review by gtaylor — APPROVED** [review]

**Comment by seanknox** [comment]
@k8s-bot e2e test this

**Comment by flah00** [comment]
@seanknox @gtaylor 

Here's an interesting issue, the default values are supposed to allow a user to install the chart... But the `ENTRYPOINT` of the image exits, if the env var `API_KEY` isn't set. I'm inclined to say that the `datadog.apiKey` should default an empty string and the check in `NOTES.txt` should be modified to account for that possibility.

**Comment by seanknox** [comment]
FWIW here's an example of handling a lack of API key: https://github.com/kubernetes/charts/blob/master/stable/sumokube/templates/NOTES.txt

**Comment by flah00** [comment]
@seanknox The `NOTES.txt` for datadog already includes the same notification. The tests are affected by this issue, because the image's `ENTRYPOINT` will exit, if `API_KEY` isn't set. 

AFAICT, the test failure is due to the image's enforcement of `API_KEY` being set. So, we have to decide how to work around the lack of a reasonable default. 

https://github.com/DataDog/docker-dd-agent/blob/master/entrypoint.sh#L14

**Comment by flah00** [comment]
@seanknox Sometimes I am thick. I now realize, my assumption was incorrect... and your suggestion was aimed at setting a default of `MISSING`, when the user doesn't provide any value. Derp. I'll look into why this default value isn't working.

**Comment by seanknox** [comment]
@lachie83 @prydonius @mgoodness I think CI failed because the chart output has the word "error"—can one of you confirm?

**Comment by mgoodness** [comment]
I think it's just highlighting lines with the words "error" and "fail" on them. Looks to me like the build job isn't able to find/delete created resources, so it's failing. Let's try again

@k8s-bot ok to test

**Comment by prydonius** [comment]
The deletion error looks like an issue in Helm. Can anyone reproduce this locally?

**Comment by seanknox** [comment]
Aha, it's the `---` delimiters at the top of the configmaps. helm still deletes the resources despite the errors. Remove those and the deletion error goes away. @flah00 
```
---  
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ template "confd.fullname" . }}
  labels:
    app: {{ template "confd.fullname" . }}
    chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
    release: "{{ .Release.Name }}"
    heritage: "{{ .Release.Service }}"
data:
{{- if .Values.datadog.confd }}
{{ toYaml .Values.datadog.confd | indent 2 }}
{{- end -}}
```

File: stable/datadog/templates/_helpers.tpl [hunk 1]
```diff
@@ -3,14 +3,39 @@
...
+{{- define "confd.fullname" -}}
+{{- printf "%s-datadog-confd" .Release.Name | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
+
+{{/*
+Create a default fully qualified autoconf name.
+We truncate at 63 chars because some Kubernetes name fields are limited to this (by the DNS naming spec).
+*/}}
+{{- define "autoconf.fullname" -}}
+{{- printf "%s-datadog-autoconf" .Release.Name | trunc 22 | trimSuffix "-" -}}
```
Comment (seanknox) [inline stable/datadog/templates/_helpers.tpl]: this should be `trunc 63`

File: stable/datadog/templates/autoconf-configmap.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+---
```
Comment (seanknox) [inline stable/datadog/templates/autoconf-configmap.yaml]: The YAML delimiters at the top of the configmaps trip up `helm delete` for some reason (and CI fails). You'll need to remove them.

**Review by seanknox — DISMISSED** [review]

**Comment by flah00** [comment]
Thanks for the help on this! Changes made.

**Comment by fiunchinho** [comment]
I'd love to see this merged!

**Comment by lachie83** [comment]
@seanknox are we clear to merge this?

**Comment by seanknox** [comment]
@lachie83 yep this is ready to go.

**Pull request merged**